### ISSUE TITLE
Bug: [no-unnecessary-template-expression] should report unnecessary template string in type context

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#9971]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&fileType=.tsx&code=C4TwDgpgBMELZgDYENYGVgCcCWA7A5gCrjQC8UABgGbIAkA3gOQqMC%2BVEiFA3EA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK18iALbF4QifQAexaIMocm6KAHcRG8GAC%2BIA0A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type templateStringType = `fa${'la'}fel`;
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-template-expression": "warn"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

```ts
type templateStringType = `fa${'la'}fel'`
//                             ~~~~  [no-unnecessary-template-expression]: Template literal expression is unnecessary and can be simplified.
```

### Actual Result

No error

**Comment by JoshuaKGoldberg** [comment]
🧆
